File: packages/typescript/src/agt-audit.ts
import {createHash} from "node:crypto";
import {EventFactory} from "./factory.js";
import type {NormalizedEvent} from "./types.js";

type Source = Record<string, unknown>; type Digest = {algorithm: string; value: string};
const NAMESPACE = "398428fd-2730-498f-8ed8-6b4290668171";

export function agtAuditPolicyDecision(factory: EventFactory, source: Source, options: {runId: string; policyEngineVersion: string; bundleDigest: Digest; resourceType: string; evaluationDurationNs?: number; enforcementMode?: string}): NormalizedEvent {
  const sourceType = required(source.event_type, "event_type"); if (!new Set(["policy_evaluation", "policy_violation"]).has(sourceType)) throw new Error(`AGT audit event is not a policy event: ${sourceType}`);
  const duration = options.evaluationDurationNs ?? 0; safeDuration(duration, "evaluationDurationNs");
  return factory.build("policy.decision", {runId: options.runId, agentId: required(source.agent_did, "agent_did"), eventId: sourceEventId("policy", required(source.entry_id, "entry_id")), timeUnixNano: timestampNs(source.timestamp, "timestamp"), ...(source.trace_id === undefined || source.trace_id === null ? {} : {traceId: required(source.trace_id, "trace_id")})}, {
    decision: policyDecision(source.policy_decision), policy: {engine: "agt", engine_version: required(options.policyEngineVersion, "policyEngineVersion"), bundle_digest: options.bundleDigest, ...(source.matched_rule === undefined || source.matched_rule === null ? {} : {policy_id: required(source.matched_rule, "matched_rule")})}, action_type: required(source.action, "action"), resource_type: required(options.resourceType, "resourceType"), enforcement_mode: options.enforcementMode ?? "enforce", evaluation_duration_ns: duration, reason_codes: [`agt.audit:${sourceType}`],
  });
}

export function agtAuditAction(factory: EventFactory, source: Source, options: {runId: string; actionDigest: Digest; actionKind: string; operation: string; durationNs?: number}): NormalizedEvent {
  const sourceType = required(source.event_type, "event_type"); if (!new Set(["tool_invocation", "tool_blocked", "action"]).has(sourceType)) throw new Error(`AGT audit event is not an action event: ${sourceType}`);
  const duration = options.durationNs ?? auditDuration(source); safeDuration(duration, "durationNs");
  const entryId = required(source.entry_id, "entry_id");
  return factory.build("action.executed", {runId: options.runId, agentId: required(source.agent_did, "agent_did"), eventId: sourceEventId("action", entryId), timeUnixNano: timestampNs(source.timestamp, "timestamp"), ...(source.trace_id === undefined || source.trace_id === null ? {} : {traceId: required(source.trace_id, "trace_id")})}, {
    action_id: entryId, action_kind: options.actionKind, action_name: required(source.action, "action"), operation: required(options.operation, "operation"), outcome: sourceType === "tool_blocked" ? "denied" : actionOutcome(source.outcome), duration_ns: duration, action_digest: options.actionDigest, ...(source.target_did === undefined || source.target_did === null ? {} : {target: {kind: "agent", id: required(source.target_did, "target_did")}}),
  });
}

function policyDecision(value: unknown): string { const result = new Map<unknown, string>([["allow", "allow"], ["allowed", "allow"], ["deny", "deny"], ["denied", "deny"], ["require_approval", "challenge"], ["requires_approval", "challenge"], ["review", "challenge"], ["not_applicable", "not_applicable"], ["error", "error"]]).get(value); if (!result) throw new Error(`unsupported AGT audit policy decision: ${String(value)}`); return result; }
function actionOutcome(value: unknown): string { const result = new Map<unknown, string>([["success", "success"], ["failure", "error"], ["error", "error"], ["denied", "denied"], ["cancelled", "cancelled"], ["timeout", "timeout"]]).get(value); if (!result) throw new Error(`unsupported AGT audit action outcome: ${String(value)}`); return result; }
function auditDuration(source: Source): number { if (source.issued_at === undefined || source.issued_at === null || source.completed_at === undefined || source.completed_at === null) throw new Error("AGT action audit requires durationNs or both issued_at and completed_at"); const start = timestampNs(source.issued_at, "issued_at"); const end = timestampNs(source.completed_at, "completed_at"); if (end < start) throw new Error("AGT completed_at cannot predate issued_at"); const value = end - start; if (value > BigInt(Number.MAX_SAFE_INTEGER)) throw new Error("AGT audit duration exceeds the safe integer range"); return Number(value); }
function safeDuration(value: number, field: string): void { if (!Number.isSafeInteger(value) || value < 0) throw new Error(`${field} must be a non-negative safe integer`); }
function required(value: unknown, field: string): string { if (typeof value !== "string" || !value) throw new Error(`AGT audit ${field} must be a non-empty string`); return value; }
function timestampNs(value: unknown, field: string): bigint { if (typeof value !== "string") throw new Error(`AGT audit ${field} must be an RFC 3339 UTC timestamp`); const match = /^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})(?:\.(\d{1,9}))?(?:Z|\+00:00)$/.exec(value); if (!match) throw new Error(`AGT audit ${field} must be an RFC 3339 UTC timestamp`); const milliseconds = Date.parse(`${match[1]}Z`); if (match[1]!.startsWith("0000") || !Number.isFinite(milliseconds) || new Date(milliseconds).toISOString().slice(0, 19) !== match[1]) throw new Error(`AGT audit ${field} is invalid`); return BigInt(milliseconds) * 1_000_000n + BigInt((match[2] ?? "").padEnd(9, "0") || "0"); }
function sourceEventId(kind: string, id: string): string { const ns = Buffer.from(NAMESPACE.replaceAll("-", ""), "hex"); const bytes = createHash("sha1").update(ns).update(`${kind}:${id}`, "utf8").digest().subarray(0, 16); bytes[6] = (bytes[6]! & 0x0f) | 0x50; bytes[8] = (bytes[8]! & 0x3f) | 0x80; const hex = bytes.toString("hex"); return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`; }

File: src/agentrust_telemetry/adapters/agt_audit.py
"""Strict adapters for Agent Mesh audit entries shipped by AGT core."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Any

from .base import EventFactory


_AUDIT_NAMESPACE = uuid.UUID("398428fd-2730-498f-8ed8-6b4290668171")
_POLICY_EVENTS = {"policy_evaluation", "policy_violation"}
_ACTION_EVENTS = {"tool_invocation", "tool_blocked", "action"}


def agt_audit_policy_decision(
    factory: EventFactory,
    source: Any,
    *,
    run_id: str,
    policy_engine_version: str,
    bundle_digest: dict[str, str],
    resource_type: str,
    evaluation_duration_ns: int = 0,
    enforcement_mode: str = "enforce",
) -> dict[str, Any]:
    """Map a policy-oriented Agent Mesh AuditEntry without copying its data."""
    event_type = _required_string(_field(source, "event_type"), "event_type")
    if event_type not in _POLICY_EVENTS:
        raise ValueError(f"AGT audit event is not a policy event: {event_type!r}")
    decision = _policy_decision(_field(source, "policy_decision"))
    entry_id = _required_string(_field(source, "entry_id"), "entry_id")
    policy: dict[str, Any] = {
        "engine": "agt",
        "engine_version": _required_string(policy_engine_version, "policy_engine_version"),
        "bundle_digest": bundle_digest,
    }
    matched_rule = _field(source, "matched_rule")
    if matched_rule is not None:
        policy["policy_id"] = _required_string(matched_rule, "matched_rule")
    return factory.build(
        "policy.decision",
        run_id=run_id,
        agent_id=_required_string(_field(source, "agent_did"), "agent_did"),
        event_id=_source_event_id("policy", entry_id),
        time_unix_nano=_datetime_ns(_field(source, "timestamp"), "timestamp"),
        trace_id=_optional_string(_field(source, "trace_id"), "trace_id"),
        decision=decision,
        policy=policy,
        action_type=_required_string(_field(source, "action"), "action"),
        resource_type=_required_string(resource_type, "resource_type"),
        enforcement_mode=enforcement_mode,
        evaluation_duration_ns=evaluation_duration_ns,
        reason_codes=[f"agt.audit:{event_type}"],
    )


def agt_audit_action(
    factory: EventFactory,
    source: Any,
    *,
    run_id: str,
    action_digest: dict[str, str],
    action_kind: str,
    operation: str,
    duration_ns: int | None = None,
) -> dict[str, Any]:
    """Map an action audit row using a caller-computed full action digest."""
    event_type = _required_string(_field(source, "event_type"), "event_type")
    if event_type not in _ACTION_EVENTS:
        raise ValueError(f"AGT audit event is not an action event: {event_type!r}")
    entry_id = _required_string(_field(source, "entry_id"), "entry_id")
    outcome = "denied" if event_type == "tool_blocked" else _action_outcome(
        _field(source, "outcome")
    )
    resolved_duration = duration_ns
    if resolved_duration is None:
        resolved_duration = _audit_duration(source)
    target: dict[str, str] | None = None
    target_did = _field(source, "target_did")
    if target_did is not None:
        target = {"kind": "agent", "id": _required_string(target_did, "target_did")}
    return factory.build(
        "action.executed",
        run_id=run_id,
        agent_id=_required_string(_field(source, "agent_did"), "agent_did"),
        event_id=_source_event_id("action", entry_id),
        time_unix_nano=_datetime_ns(_field(source, "timestamp"), "timestamp"),
        trace_id=_optional_string(_field(source, "trace_id"), "trace_id"),
        action_id=entry_id,
        action_kind=action_kind,
        action_name=_required_string(_field(source, "action"), "action"),
        operation=_required_string(operation, "operation"),
        outcome=outcome,
        duration_ns=resolved_duration,
        action_digest=action_digest,
        **({"target": target} if target else {}),
    )


def _policy_decision(value: Any) -> str:
    mapping = {
        "allow": "allow",
        "allowed": "allow",
        "deny": "deny",
        "denied": "deny",
        "require_approval": "challenge",
        "requires_approval": "challenge",
        "review": "challenge",
        "not_applicable": "not_applicable",
        "error": "error",
    }
    if value not in mapping:
        raise ValueError(f"unsupported AGT audit policy decision: {value!r}")
    return mapping[value]


def _action_outcome(value: Any) -> str:
    mapping = {
        "success": "success",
        "failure": "error",
        "error": "error",
        "denied": "denied",
        "cancelled": "cancelled",
        "timeout": "timeout",
    }
    if value not in mapping:
        raise ValueError(f"unsupported AGT audit action outcome: {value!r}")
    return mapping[value]


def _audit_duration(source: Any) -> int:
    issued = _field(source, "issued_at")
    completed = _field(source, "completed_at")
    if issued is None or completed is None:
        raise ValueError(
            "AGT action audit requires duration_ns or both issued_at and completed_at"
        )
    issued_ns = _datetime_ns(issued, "issued_at")
    completed_ns = _datetime_ns(completed, "completed_at")
    if completed_ns < issued_ns:
        raise ValueError("AGT completed_at cannot predate issued_at")
    return completed_ns - issued_ns


def _source_event_id(kind: str, source_id: str) -> str:
    return str(uuid.uuid5(_AUDIT_NAMESPACE, f"{kind}:{source_id}"))


def _field(source: Any, name: str) -> Any:
    if isinstance(source, dict):
        return source.get(name)
    return getattr(source, name, None)


def _required_string(value: Any, field: str) -> str:
    if not isinstance(value, str) or not value:
        raise ValueError(f"AGT audit {field} must be a non-empty string")
    return value


def _optional_string(value: Any, field: str) -> str | None:
    return None if value is None else _required_string(value, field)


def _datetime_ns(value: Any, field: str) -> int:
    if not isinstance(value, datetime) or value.tzinfo is None:
        raise ValueError(f"AGT audit {field} must be a timezone-aware datetime")
    utc = value.astimezone(timezone.utc)
    epoch = datetime(1970, 1, 1, tzinfo=timezone.utc)
    delta = utc - epoch
    return (
        delta.days * 86_400_000_000_000
        + delta.seconds * 1_000_000_000
        + delta.microseconds * 1_000
    )
